### PR TITLE
Globals

### PR DESCRIPTION
**Pull request by pbrunet**
This one is ready for review.

* static variable for globals look useless as constant folding already expend these values.
* local_decl and local_declaration looks really similare. I didn't merge them because
they don't return the same type. We should be able to return only strings but it can't be done without modification in typing (and I am not sure it is a good idea...). If we return only ast.Name, we have to create dummy ast.Name (for function name for example) and it will also need modification in typing.
We could return a dict with both string_name and ast.Name but typing will have to handle dummy ast.Name (once again, we have to touch typing)

### CODE REVIEW

File: doc/INTERNAL.rst [hunk 1]
```diff
@@ -200,27 +200,28 @@ show that Pythran can compile it::
...
 
 While statement at top-level::
 
-    $> printf 'i=0\nwhile i<10:i+=1\nprint i' > internal_top_level_while.py
+    $> printf 'i=0\nwhile True:\n print i\n break\nprint i' > internal_top_level_while.py
     $> pythran internal_top_level_while.py
     $> python -c 'import internal_top_level_while'
-    10
+    0
+    0
```
Comment (serge-sans-paille) [inline doc/INTERNAL.rst]: why did you change these examples? The new version do not look that nice, especially the break in the while.

File: pythran/analyses/global_declarations.py [hunk 2]
```diff
@@ -1,22 +1,46 @@
...
+    ... from math import cos
+    ... c = 12
+    ... def foo(a):
+    ...     b = a + 1''')
+    >>> pm = passmanager.PassManager("test")
+    >>> pm.gather(GlobalDeclarations, node).keys()
+    ['cos', 'foo', 'c', 'math']
+
+    """
+
```
Comment (serge-sans-paille) [inline pythran/analyses/global_declarations.py]: I really appreciate the documentation effort!

File: pythran/analyses/global_declarations.py [hunk 3]
```diff
@@ -1,22 +1,46 @@
...
+        """ Imported functions define a new variable name. """
         self.result.update((a.asname or a.name, a) for a in node.names)
 
     def visit_FunctionDef(self, node):
+        """ Function definition defined a new identifier. """
         self.result[node.name] = node
         # no generic visit here, so no diving into function body
+
+    def visit_Assign(self, node):
+        """ Assigned global variable define a new identifier. """
```
Comment (serge-sans-paille) [inline pythran/analyses/global_declarations.py]: an assert isinstance ast.Name would not hurt here. You could use dict.update to avoid the explicit loop ;-)

File: doc/INTERNAL.rst (hunk 1 above)
Comment (pbrunet) [inline doc/INTERNAL.rst]: Because i += 1 is a modification on globals which is not allowed anymore.

File: pythran/backend.py [hunk 4]
```diff
@@ -573,7 +573,12 @@ def visit_Return(self, node):
                     Cxx.final_statement))
                 ])
         else:
-            stmt = ReturnStatement(self.visit(node.value))
+            value = self.visit(node.value)
+            if metadata.get(node, metadata.StaticReturn):
+                stmt = Block([Assign("static auto tmp_global", value),
+                              ReturnStatement("tmp_global")])
+            else:
+                stmt = ReturnStatement(value)
```
Comment (serge-sans-paille) [inline pythran/backend.py]: What about adding a static qualifier to the type, and not modifying the back-end?

File: pythran/cxxtypes.py [hunk 5]
```diff
@@ -40,7 +40,8 @@ def all_types(self):
     def __eq__(self, other):
         havesameclass = self.__class__ == other.__class__
         if havesameclass:
-            def same(x, y): return getattr(self, x) == getattr(other, y)
+            def same(x, y):
+                return getattr(self, x) == getattr(other, y)
```
Comment (serge-sans-paille) [inline pythran/cxxtypes.py]: Fixed in the decl-impl branch?

File: pythran/syntax.py [hunk 6]
```diff
@@ -46,18 +46,6 @@ def save_attribute(module):
...
-               " or imports")
-        for n in node.body:
-            if isinstance(n, ast.Expr) and isinstance(n.value, ast.Str):
-                continue
-            else:
-                if not any(isinstance(n, getattr(ast, t))
-                           for t in ('FunctionDef', 'Import', 'ImportFrom',)):
-                    raise PythranSyntaxError(err, n)
-        self.generic_visit(node)
-
```
Comment (serge-sans-paille) [inline pythran/syntax.py]: check we only have top-level assign to name?

File: pythran/transformations/expand_globals.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    def visit_Module(self, node):
+        """Turn globals assignement to functionDef and visit function defs. """
+        module_body = list()
+        # Gather top level assigned variables.
+        for stmt in node.body:
+            if not isinstance(stmt, ast.Assign):
+                continue
+            for target in stmt.targets:
+                assert isinstance(target, ast.Name)
```
Comment (serge-sans-paille) [inline pythran/transformations/expand_globals.py]: this looks like an analyse to me, you should filter out Function from the global decl instead?

File: pythran/transformations/expand_globals.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                cst_value = self.visit(stmt.value)
+                for target in stmt.targets:
+                    assert isinstance(target, ast.Name)
+                    module_body.append(
+                        ast.FunctionDef(target.id,
+                                        ast.arguments([], None, None, []),
+                                        [ast.Return(value=cst_value)], []))
+                    metadata.add(module_body[-1].body[0], metadata.StaticReturn())
+            else:
+                self.local_decl = self.passmanager.gather(LocalDecl, stmt,
```
Comment (serge-sans-paille) [inline pythran/transformations/expand_globals.py]: I don't understand this compared to using an existing analyse either :-/

File: pythran/cxxtypes.py (hunk 5 above)
Comment (pbrunet) [inline pythran/cxxtypes.py]: Yes it is. (I fixed it here too to use the failfast arg on my computer :) )

File: pythran/transformations/expand_globals.py (hunk 7 above)
Comment (pbrunet) [inline pythran/transformations/expand_globals.py]: Not only, I should filter, import and functions (indeed, everything except assignement to ast.Name)

I agree it looks like an analyse. I was just not sure we should create an analyse for this "very simple" case.

File: pythran/backend.py (hunk 4 above)
Comment (pbrunet) [inline pythran/backend.py]: May be. But I really wonder if this commit should be done as generated code is never used...

File: pythran/transformations/expand_globals.py (hunk 7 above)
Comment (serge-sans-paille) [inline pythran/transformations/expand_globals.py]: > Not only, I should filter, import and functions (indeed, everything except assignement to ast.Name)

So you have a very simple filter here, based on the xisting
globaldeclaration?

File: doc/INTERNAL.rst (hunk 1 above)
Comment (serge-sans-paille) [inline doc/INTERNAL.rst]: > Because i += 1 is a modification on globals which is not allowed anymore.
ok. Should we put this in a functiondef instead of top-level, for the
sake of the doc?

File: pythran/transformations/expand_globals.py (hunk 8 above)
Comment (pbrunet) [inline pythran/transformations/expand_globals.py]: I don't understand what you don't understand.

File: doc/INTERNAL.rst (hunk 1 above)
Comment (pbrunet) [inline doc/INTERNAL.rst]: No because we want to test the __init__ function for top level stmt. If we put it in a function def, it doesn't make sens.

File: pythran/transformations/expand_globals.py (hunk 7 above)
Comment (pbrunet) [inline pythran/transformations/expand_globals.py]: Yes, it could be done that way.

File: pythran/backend.py (hunk 4 above)
Comment (serge-sans-paille) [inline pythran/backend.py]: > May be. But I really wonder if this commit should be done as generated code is never used...
??

File: pythran/transformations/expand_globals.py (hunk 8 above)
Comment (serge-sans-paille) [inline pythran/transformations/expand_globals.py]: > > +                self.local_decl = self.passmanager.gather(LocalDecl, stmt,
> 
> I don't understand what you don't understand.
I missed thegather call. forget about it.

File: doc/INTERNAL.rst (hunk 1 above)
Comment (serge-sans-paille) [inline doc/INTERNAL.rst]: > No because we want to test the __init__ function for top level stmt. If we put it in a function def, it doesn't make sens.
ok

File: pythran/backend.py (hunk 4 above)
Comment (pbrunet) [inline pythran/backend.py]: I did the modification to have  static value but while I was looking to the code, I notice that this code is always constant folded so this function is never used.

Comment (serge-sans-paille) [inline pythran/backend.py]: > I did the modification to have  static value but while I was looking to the code, I notice that this code is always constant folded so this function is never used.

Even when 

```python
a = [1,2,3,4]

def foo(i):
    return a[i]
```

?

Comment (pbrunet) [inline pythran/backend.py]: Yes, just like with

```Python
def foo():
    return [1, 2, 3]

def bar(i):
    print foo()[i]
```

**Comment by serge-sans-paille** [comment]
@pbrunet I wonder what's the status on this one ^^!

**Comment by pbrunet** [comment]
I will not have time for the implementation using real globals instead of function.
I think we can:
1) I fix this one with your current comment
2) You merge it even if it is not the most efficient
3) Someone code a better globals implementation. (Now? Or after the better typing?)

Do you agree?

**Comment by serge-sans-paille** [comment]
On Sun, Jun 14, 2015 at 10:43:18AM -0700, pbrunet wrote:
> I will not have time for the implementation using real globals instead of function.
> I think we can:
> 1) I fix this one with your current comment
> 2) You merge it even if it is not the most efficient
> 3) Someone code a better globals implementation. (Now? Or after the better typing?)

If 1) includes the fix for the static storage, I agree with it.
I dislike 2) because the performance issue does not look minor to me.
I prefer 1) over 3)

:-)

**Comment by pbrunet** [comment]
As we said, the static storage is useless because of constant folding for this function. (1, 2, 3 were more steps than choices :) )

**Comment by serge-sans-paille** [comment]
Can't we prevent constant folding in that case?

```
a = [1,2,...,100000]

def foo(i):
    return a[i]
```

As far as I understand, it is turned into

```
def global_a():
    return [1,2,...,100000]

def foo(i):
    return [1,2,...,100000][i]
```

Which seems a common case to me, and very bad for performance.

(sorry if history repeats again, that's an undesirable property of
history :-/)

**Comment by pbrunet** [comment]
According to the constant folding rules, this list is "too big to be constant folded" so it will work.

IMO, It doesn't look to be a common case. most of the time, const globals variables are : scalar (PI, EARTH_RADIUS, CELERITY, ....) or dict of value (to group variable by interest).

**Comment by serge-sans-paille** [comment]
@pbrunet maybe I could rework this if you allow me to?

**Comment by pbrunet** [comment]
I just don't know the status of this one. I don't know what should be done and what was wrong.

**Comment by serge-sans-paille** [comment]
On Sun, Dec 27, 2015 at 04:57:03AM -0800, pbrunet wrote:
> I just don't know the status of this one. I don't know what should be done and what was wrong.

If I recall correctly, I disliked (and still do) the fact that you
convert


a = [1, 2, 3,....., 10000]

def foo(i):
    return a[i]

into

def a():
    return [1, 2, 3,....., 10000]

def foo(i):
    return a()[i]

because this would lead to a multiple creattion of a big array.

The solution would be to generate

def a(l=[1, 2, 3,....., 10000]):
    return l

but this is not correctly converted by pythran, so let this be my next
task: correctly handle default parameter when it's an object!

**Comment by pbrunet** [comment]
In fact, I changed the backend to generate static variables so the
performance penalty is not a problem and there are not multiple creation of
big array.

2015-12-27 18:51 GMT+01:00 serge-sans-paille <notifications@github.com>:

**Comment by serge-sans-paille** [comment]
On Mon, Dec 28, 2015 at 03:34:13AM -0800, pbrunet wrote:
> In fact, I changed the backend to generate static variables so the
> performance penalty is not a problem and there are not multiple creation of
> big array.
ok, I'll review that (there seems to need a rebase though)
